AO3-5556 Don't change collections on preview.

tickinginstant opened the pull request:
## Issue

https://otwarchive.atlassian.net/browse/AO3-5556

## Purpose

- Generalize the function `assign_tags_of_type` to a function `assign_through_association` defined in `ApplicationRecord` that can be used for any type of association, not just tags.
- Use that function to assign collections, so that setting `collection_names` will only set up the changes to be performed on save instead of performing them immediately. Similarly with `collections_to_add` and `collections_to_remove`.
- Change the `approve_automatically` and `set_anonymous_and_unrevealed` callbacks in the `CollectionItem` class to run `before_validation` instead of `before_save`, so that we get the correct values for `anonymous`, `unrevealed`, `user_approval_status`, and `collection_approval_status` after validation.
- Make the `Collectible` module a concern and move it to the correct directory.
- Rewrite the `Collectible` function `set_anon_unrevealed` so that it uses the status of the collection items to compute whether the work should be anonymous/unrevealed, and make it run as an `after_validation` callback so that it has access to the collection item statuses set before validation.
- Rewrite `BookmarksController#update` to use the collection handling from `Collectible`, and add a function akin to `WorksController#in_moderated_collection` so that the message on adding a moderated collection stays the same as it was before the change.
- Stop calling `save(validate: false)` so much, and replace it with a `reindex_item` callback in the `CollectionItem` class. The `item` association on `CollectionItem`s has `touch: true`, so changes to collection items will already update the work/bookmark timestamp. And indexing is the only other real purpose that calling `save(validate: false)` serves.

tickinginstant committed:
AO3-5556 Don't change collections on preview.

tickinginstant committed:
AO3-5556 For Hound.

redsummernight commented on app/models/application_record.rb:
```diff
@@ -15,4 +15,35 @@ def update_sanitizer_version
     order(Arel.sql("RAND()"))
   end
+
+  def unmark_for_destruction
```
Can you add a simple test that calls `mark_for_destruction` then `unmark_for_destruction` on a record, and expects `marked_for_destruction?` to be false?

sarken replied:
Just checking -- is this ready now that the tests have been added, or does it need a more thorough review?

tickinginstant committed:
AO3-5556 Add test for unmark_for_destruction.

tickinginstant committed:
Merge branch 'master' into AO3-5556-collection-preview

tickinginstant committed:
AO3-5556 Fix bad merge.

tickinginstant committed:
AO3-5556 Better fix.

tickinginstant committed:
AO3-5556 Fix intermittent test failure.

I think it's pre-existing, but it's vaguely related, so I might as well
fix it.

tickinginstant committed:
Merge branch 'main' into AO3-5556-collection-preview

tickinginstant committed:
Merge branch 'master' into AO3-5556-collection-preview

pmonfort commented on app/models/work.rb:
```diff
@@ -569,17 +520,8 @@ def visible?(user = User.current_user)
-    #!self.collection_items.anonymous.empty?
-    in_anon_collection?
-  end
+  alias_attribute :unrevealed, :in_unrevealed_collection
```
👏

pmonfort commented on app/models/collection.rb:
```diff
@@ -321,65 +324,30 @@ def gift_notification
-  def unrevealed?() = self.collection_preference.unrevealed
-
-  def anonymous?() = self.collection_preference.anonymous
+  delegate :moderated?,
```
💪

pmonfort commented on app/views/works/_meta.html.erb:
```diff
@@ -49,12 +49,13 @@
     <% end %>
 
-    <% unless @work.approved_collections.empty? %>
+    <% collections_data = show_collections_data(@work.approved_collections_after_saving) %>
```
Non-blocking suggestion: I'm not a big fan of setting variables in views. One alternative would be to set this in the controller, but since this is a partial that could be rendered from different contexts, I think a helper would be cleaner

```
# app/helpers/works_helper.rb
def collections_meta_tag(work)
  collections_data = show_collections_data(work.approved_collections_after_saving)
  return if collections_data.blank?
  
  content_tag(:dt, ts("Collections:"), class: "collections") +
    content_tag(:dd, collections_data, class: "collections")
end
```

Then the view becomes:

<%= collections_meta_tag(@work) %>

pmonfort commented on app/models/application_record.rb:
```diff
@@ -15,4 +15,35 @@ def update_sanitizer_version
     order(Arel.sql("RAND()"))
   end
+
+  def unmark_for_destruction
```
Both `unmark_for_destruction` and `assign_through_association` are only used by `Collectible` and `Taggable` should they live in ApplicationRecord? Maybe they can be extracted to a concern.

pmonfort commented on app/models/application_record.rb:
```diff
@@ -15,4 +15,35 @@ def update_sanitizer_version
+  # The optional parameter klass is used to filter the through records -- when
+  # it's set, this function won't mark or unmark any through records whose
+  # source doesn't have that klass.
+  def assign_through_association(through_association, source, values, klass: nil)
```
See comment on unmark_for_destruction above, same concern about scope in ApplicationRecord

pmonfort reviewed (commented):
Great work, this PR covers a lot of ground. I left some comments, let me know your thoughts.

pmonfort commented:
Hi! I reviewed this PR and left some comments. As discussed with the team, I'll be opening a follow-up PR based on your work to resolve the feedback and merge conflicts. Thanks!
